### ISSUE TITLE
Release 14.0.0

### ISSUE THREAD

**Issue by jeddy3** [stylelint/stylelint#5205]
Node 10 EOL is at the end of April. A lot of the packages we depend on will shift to Node 12 (and ESM) around then (see https://github.com/stylelint/stylelint/pull/5198 and https://github.com/sindresorhus/meta/discussions/15)

We should shoot to migrate stylelint to ESM for May. It'll be a major release of stylelint, so it's an opportunity to roll in other breaking changes. If you've got an itch to scratch that's a breaking change, then this is the time to do it!

We should remove deprecated code, e.g. https://github.com/stylelint/stylelint/issues/4899 and all the deprecated rules.

**Comment by ybiquitous** [comment]
If we can use [dynamic imports](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/import#dynamic_imports), `import-lazy` may be needless:

https://github.com/stylelint/stylelint/blob/e25b5a203deb63c642e0b3b4bbab8a320180725c/package.json#L128

**Comment by jeddy3** [comment]
Does node@12 supports dynamic imports? If so, that'll be great as we are reliant lazy/dynamic loading for performance.

**Comment by ybiquitous** [comment]
Oh, I missed that the `--experimental-modules` flag is necessary for Node 12, sorry 😓 

<img width="785" alt="image" src="https://user-images.githubusercontent.com/473530/114325435-21742d80-9b6b-11eb-8cac-76cdf7fd363c.png">

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/import#browser_compatibility

**Comment by m-allanson** [comment]
I think the latest LTS release of node 12 supports ESM without a flag.  See:

- https://github.com/nodejs/node/pull/37718/files
- https://nodejs.org/docs/latest-v12.x/api/esm.html#esm_modules_ecmascript_modules

<img width="636" alt="Screenshot 2021-04-12 at 08 46 35" src="https://user-images.githubusercontent.com/381801/114359250-ab23fb00-9b6b-11eb-9cdc-7b3c50f62947.png">

**Comment by hudochenkov** [comment]
@m-allanson I think we talk about `import().then()`, not regular `import`.

**Comment by jeddy3** [comment]
> Oh, I missed that the --experimental-modules flag is necessary for Node 12, sorry

Perhaps ESM (non-dynamic) with bundling will be our approach for 14.0.0. It'll give us:

- something that's future-facing (for dynamic ESM down the line)
- compatibility with our dependencies which will be shifting to ESM
- a non-performant version that can be installed directly from GitHub and tested against
- a performant version published to NPM with a small set of system tests
- browser compatibility

Steps towards this:

1. Deprecated and then remove problematic `function-calc-no-invalid` (https://github.com/stylelint/stylelint/issues/4731#issuecomment-623440469).
2. Investigate if there's a way to refactor the `no-vendor` prefix rules to avoid relying on autoprefixer (and the caniuse dataset); prefixes are on their way out in favour of browser flags, and perhaps maintaining a simple list in stylelint of vendor-prefixed properties used today is a simpler viable alternative.
3. Investigate ESM bundlers and compare their performance, particular ESBuild.
4. Investigate Jest's native support for ESM (we'll want to avoid transpiling for tests).

**Comment by alexander-akait** [comment]
> Investigate Jest's native support for ESM (we'll want to avoid transpiling for tests).

Not ready yet, and I think will not ready in near future

**Comment by ybiquitous** [comment]
FYI. https://jestjs.io/docs/ecmascript-modules

It seems Jest's mocks do not work with ESM: https://github.com/facebook/jest/issues/10025

**Comment by jeddy3** [comment]
Thoughts on a node 14 minimum in `14.0.0`?

We're such a small team of volunteers that I think we may need to jump to a minimum of node 14 if we are to stand a chance of migrating to ESM to maintain compatibility with our dependencies. As a dev tool, I think we can be progressive in the node versions we support.

We'd then be able to use dynamic imports for rules, syntaxes, formatters, plugins and processors, which will hopefully keep things simple and doable. If anyone fancies it, a proof of concept showing dynamic imports working (for just one of these parts) would be ace.

(It'd also be good to know if [either method B or C](https://bl.ocks.org/rstacruz/511f43265de4939f6ca729a3df7b001c) will work with transpiling dynamic imports and Jest.).

**Comment by ybiquitous** [comment]
The idea sounds attractive to me. But the one thing that I’m concerned about is that perhaps many people will still be using Node.js 12 until it will be [End-Of-Life on 2022-04-30](https://github.com/nodejs/Release#release-schedule).

For example, the following application will not be able to install stylelint 14:

- it has `"engines": { "node": "12.22.0" }` in its `package.json`
- and, it has `engine-strict = true` in its `.npmrc`

Such applications must continue using stylelint 12 or upgrade to Node.js 14. The task may be hard for some applications... 🤔 

There is a more flexible but a bit complex way to use dynamic imports and require Node.js 12 as a minimum. In that case, people using Node.js 12 would have to specify the `--experimental-modules` flag such as `NODE_OPTIONS=--experimental-modules stylelint *.css`. We would describe such a workaround and promote users to upgrade to Node.js 14 in the document.

**Comment by jaydenseric** [comment]
Why is there a dilemma? With this level of support:

```json
{
  "engines": {
    "node": "^12.20 || >= 14.13"
  }
}
```

You can use everything to do with ESM in Node.js except top level await, which is available unflagged since v14.8.0:

https://nodejs.org/dist/latest-v16.x/docs/api/esm.html#esm_modules_ecmascript_modules

Conveniently, this is also matches the minimum Node.js versions that support [package subpath patterns](https://nodejs.org/dist/latest-v16.x/docs/api/packages.html#packages_subpath_patterns).

I'm migrating all of my published packages to this level of Node.js support, and so are many other package maintainers (although some are lazy and incorrectly just update `engines.node` to `>=12`).

**Comment by ybiquitous** [comment]
@jaydenseric Excellent idea! 👏🏼 It should resolve any concerns!

**Comment by jeddy3** [comment]
Thanks for clearing that up. I hadn't realised dynamic imports are available in 12.20 without a flag.

I think we're good to proceed with moving to ESM with:

```json
{
  "engines": {
    "node": "^12.20 || >= 14.13"
  }
}
```

Next steps:

- investigating dynamic imports
- investigating using Jest with ESM

If anyone fancies digging in, please do.

**Comment by qubyte** [comment]
One such package which has dropped support for node 10 is meow, which stylelint depends on. Version 9 is afflicted with [this advisory](https://www.npmjs.com/advisories/1677) in its dependency tree, which is resolved naturally in version 10. Version 10 of meow drops support for node 10 though, so updating to it would mean dropping node 10 support here too.

**Comment by jeddy3** [comment]
Thanks.

Meow [10.0.0 release notes](https://github.com/sindresorhus/meow/releases/tag/v10.0.0) and [linked ESM migration guide](https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c):

> ### How can I move my CommonJS project to ESM?

> - Add `"type": "module"` to your package.json.
> - Replace `"main": "index.js"` with `"exports": "./index.js"` in your package.json.
> - Update the `"engines"` field in package.json to Node.js 12: `"node": ">=12.20"`.
> - Remove `'use strict';` from all JavaScript files.
> - Replace all `require()`/`module.export` with `import`/`export`.
> - Use only full relative file paths for imports: `import x from '.';` → `import x from './index.js';`.
> - If you have a TypeScript type definition (for example, `index.d.ts`), update it to use ESM imports/exports.
> - Optional but recommended, use the [`node:` protocol](https://nodejs.org/api/esm.html#esm_node_imports) for imports.

**Comment by jeddy3** [comment]
I'm going to create separate issues for bits in the 14.0.0 release. I'll use this issue as an umbrella, and create an ordered checklist in the top post so that we have a clear set of steps.

**Comment by jeddy3** [comment]
I've created a [`v14` branch](https://github.com/stylelint/stylelint/tree/v14). I'm going to get the ball rolling by removing the deprecated rules (https://github.com/stylelint/stylelint/issues/5290).

**Comment by ybiquitous** [comment]
v14.0.0 may be a good chance to rename the main branch from `master` to `main`. What do you think?

**Comment by jeddy3** [comment]
Now that https://github.com/stylelint/stylelint/issues/5289 is closed, all the [checklist items](https://github.com/stylelint/stylelint/issues/5205#issue-837034702) are unblocked. 

If you'd like to help us release `14.0.0`, please consider [contributing](https://stylelint.io/contributing) by picking up any of the issues still left to do.

**Comment by jeddy3** [comment]
@hudochenkov Fantastic work on [the `Document` node in PostCSS](https://github.com/postcss/postcss/releases/tag/8.3.0)! That was one of the bigger pieces of the `14.0.0` jigsaw puzzle!

**Comment by hudochenkov** [comment]
@jeddy3 thank you! Next I'll be working on CSS-in-JS syntax (styled components flavor). I don't think I'll be working on any other syntax any time soon, because this is what I use.

**Comment by jeddy3** [comment]
> Next I'll be working on CSS-in-JS syntax (styled components flavor).

Excellent.

> I don't think I'll be working on any other syntax any time soon, because this is what I use.

That's very fair. I anticipate each syntax will be written by the people who use it. For example, I like the look of [Stitches](https://stitches.dev/) and may write a syntax for it if I use it on a project. It has a unique object structure, e.g. `variant` and `conditional` properties, that would be well suited to a tailored syntax.

**Comment by Perni1984** [comment]
@jeddy3 @hudochenkov we are using https://stitches.dev/ here in our current project and I would like to use it together with stylelint, is there some kind of reference of custom-syntax that we can use to extend it to fit the stitches one?

**Comment by jeddy3** [comment]
@Perni1984 It's fantastic that you'd like to create a `postcss-stitches` custom syntax. @hudochenkov knows more about syntaxes than I do, but I'll jot down my thoughts so that you can, if you like, get the ball rolling.

Yesterday, we [finished our migration to PostCSS@8](https://github.com/stylelint/stylelint/pull/5304), which includes the `Document` node that [@hudochenkov added](https://github.com/postcss/postcss/pull/1582). Now is the perfect time to start working on `postcss-stitches` as the `Document` node is needed for these types of syntaxes that can have multiple roots.

I believe @hudochenkov is looking into a custom syntax for styled-components, which uses template literals. A custom syntax for Stitches would be fantastic as it's object-based, and we'll end up with two reference syntaxes for the community: one for template literals and the other for objects.

Stitches is also a great first choice for an object-based syntax because:
- it avoids interpolation, which keeps things simpler
- uses `$` for tokens, which the rules built-into stylelint already ignore

There's no reference syntax to extend at the moment, unless @hudochenkov has cooked one up already? What we have is the monolithic [`postcss-css-in-js`](https://github.com/stylelint/postcss-css-in-js) package, which relies on the [`postcss-syntax`](https://github.com/stylelint/postcss-syntax)—a package that does both syntax inference and the monkey-patching of a `Document` node into PostCSS@7.

The `object-*` files in [`postcss-css-in-js` repo](https://github.com/stylelint/postcss-css-in-js) could be a good starting point for `postcss-stitches`. You'll need to take into account the particulars of Stitches, e.g. the `variants` property, as the `object-*` files will blindly convert all nested objects into CSS rulesets.

The syntax should work with any PostCSS-based tooling, and so you needn't use stylelint in your tests. For example, [`postcss-css-in-js` uses autoprefixer](https://github.com/stylelint/postcss-css-in-js/blob/master/test/css-in-js.js) in some of its tests. When you do test with stylelint though, you'll want to use the `v14` branch as this contains our migration to PostCSS@8.

PostCSS repo [has documentation](https://github.com/postcss/postcss/blob/main/docs/syntax.md) for writing syntaxes too. It references [postcss-scss](https://github.com/postcss/postcss-scss) which is the kind of standalone syntax, i.e. not reliant on `postcss-syntax`, that we're moving towards.

If you do start working on `postcss-stitches` it'd be fantastic if you could share your progress with us as we're only beginning to explore how to work with PostCSS@8 and the included `Document` node.

**Comment by jeddy3** [comment]
Now that the PostCSS@8 migration is complete, I believe the next issue to work on is:

- #4843 - Add `customSyntax` option to configuration object

Shout out if you're interested in working on this.

When we have this in place, we can reach out to the SCSS community to see if someone there will willing to put together a `stylelint-config-scss-standard` shared config that bundles the following:

- `postcss-scss` syntax
- `stylelint-scss` plugin
- `stylelint-config-standard` shared config

And then:

- configures the built-in rules for SCSS
- turns on and configures all the stylistic rules in `stylelint-scss` 

For example:

```json
{
  "customSyntax": "postcss-scss",
  "plugins": ["stylelint-scss"],
  "extends": ["stylelint-config-standard"],
  "rules": {
    "block-closing-brace-newline-after": ["always", { "ignoreAtRules": ["else"] } ],
    "scss/at-else-closing-brace-newline-after": "always-last-in-chain"
    ..
  }
}
```

**Comment by Airkro** [comment]
Will https://github.com/stylelint/stylelint/issues/4501 have a chance?

**Comment by jeddy3** [comment]
Can anyone foresee any problems if we set `v14` as the main branch until it's merged? It'll save us from having to request setting the merge branch in pull requests.

**Comment by hudochenkov** [comment]
@jeddy3 I don't see a problem. However, I would add a temporary disclaimer to the readme on v14 branch.

**Comment by jeddy3** [comment]
[Thanks to @hudochenkov](https://github.com/stylelint/stylelint/pull/5521), we now have an `overrides` property in `v14`. I believe the only task remaining before we can start the migration to ESM is adding `customSyntax` to the configuration object (https://github.com/stylelint/stylelint/issues/4843) as it touches on the engine.

The two other remaining tasks can happen after the move to ESM:
- #5128 - will possibly be fixed by the autofix of https://github.com/stylelint/stylelint/pull/5516
- #5301 - I can work on this while writing the migration guide in the changelog so that they align

I've also added #5142 to the nice-to-haves, but only `customSyntax` is essential for `14.0.0` as it'll ease the migration from `13.x` for some users.

**Comment by AndreasNasman** [comment]
Not sure if this is baked into the `stylelint-config-recommended release` part of the TODO-list above, but I just noticed that `function-calc-no-invalid` should be removed from the [`stylelint-config-recommended` rules](https://github.com/stylelint/stylelint-config-recommended/blob/master/index.js#L19) when release `v14` goes live. Otherwise, every file will report a `Unknown rule function-calc-no-invalid` error if `stylelint-config-recommended` or `stylelint-config-standard` is used in a project. 🙂

**Comment by jeddy3** [comment]
@AndreasNasman Thanks!

> not entirely sure how to synchronize/link the pull request with this issue!

We'll release new version of the configs alongside the `14.0.0` release.  It's step 6 in [our release process](https://stylelint.io/maintainer-guide/releases/). Having the pull requests ready to merge is ideal, so thanks for opening it.

**Comment by ybiquitous** [comment]
`eslint-config-stylelint@14.0.0` now supports ESM. See #5536

**Comment by kuangrs** [comment]
Will #5431 have a chance in Release14.0.0? thanks !

**Comment by jeddy3** [comment]
Thanks to @hudochenkov and @ybiquitous, I think we're now at the stage where we can assess whether migrating to ESM is viable or not. I suggest we throw something up that'll let us identify potential blockers as early as possible.

I can get the ball rolling by using `putout` to migrate to static imports so that we get an initial lay of the land.

**Comment by hudochenkov** [comment]
I would hold off with ESM for this release.

1. It will take some effort to make stylelint fully working with ESM.
2. There is a question of testing. Jest might be tricky with ESM (if it supports it).
3. Figure out dual publish ESM/CommonJS setup.
4. v14 already a huge release with lot's of breaking changes, including removing some syntaxes support.
5. We delayed this release for a long time. Nothing prevents us to release v15, when ESM is ready (if it would be even a breaking change with dual publish).

**Comment by ybiquitous** [comment]
Yes, that might be better. Currently, dual publish ESM/CJS is required, so our maintenance costs would be likely to increase.

**Comment by jeddy3** [comment]
All good points.

> It will take some effort to make stylelint fully working with ESM.

I was hoping to get a better understanding of just how much effort would be required before making a decision. However, I've been noodling with esm branch for a while now and there are still lots of unknowns.

I'll push up my branch and write up where it's at in the https://github.com/stylelint/stylelint/issues/5291 issue. Someone is welcome to pick it up, but let's knock moving to esm off the required `14.0.0` task list.

**Comment by jeddy3** [comment]
I've [reached out](https://github.com/kristerkari/stylelint-scss/issues/541) to the [stylelint-scss](https://github.com/kristerkari/stylelint-scss) maintainers in preparation for https://github.com/stylelint/stylelint/issues/5301.

Ref: https://github.com/stylelint/stylelint/issues/5205#issuecomment-856669471

**Comment by jeddy3** [comment]
I've cleaned up [task list at the top](https://github.com/stylelint/stylelint/issues/5205#issue-837034702) and added links to:
- issues in related repos like the demo and syntaxes
- pull requests in the config packages

The amount of recent progress has been fantastic! Thank you to everyone who's contributed!

I'm away and offline next week, but I hope to pick up https://github.com/stylelint/stylelint/issues/5301 when I'm back next weekend. It's currently blocked by https://github.com/kristerkari/stylelint-scss/issues/541, as we'll need to see if there's an appetite in the SCSS community for creating a `stylelint-config-standard-scss` package.

**Comment by jeddy3** [comment]
A fair bit of stuff related to the 14.0.0 release is happening outside of this repo. To make it easier to see what's being happening I'll list them here:

- [created a draft pull request for improving the migration experience (the last remaining task-item in stylelint itself)](https://github.com/stylelint/stylelint/pull/5563)
- [provided more guidance to the SCSS community on next steps](https://github.com/kristerkari/stylelint-scss/issues/541#issuecomment-927022891)
- [created an issue for the VS Code extension](https://github.com/stylelint/vscode-stylelint/issues/233)
- [created a draft pull request to prepare the stylelint online demo for version 14](https://github.com/stylelint/stylelint-demo/pull/270)

I'm away and offline again from Tuesday for another week, but I intend to continue with and address any feedback on https://github.com/stylelint/stylelint/pull/5563 when I get back.

Feel free to pick up https://github.com/stylelint/stylelint-demo/pull/270 in my absence, if you fancy it. That'll be really helpful. As will feeding back on https://github.com/stylelint/stylelint/pull/5563.  If anyone is feeling particularly generous, looking into https://github.com/stylelint/postcss-html/issues/2 would be fantastic especially if you lint styles in your HTML. Although we can release without it, it would be nice to have.

**Comment by jeddy3** [comment]
> Can we add #5580 to v14?

Absolutely. We won't release until we get the nod from you that you're ready. We're still getting our ducks in a row (and I'll write it a status update tomorrow).

Other people on the team are more proficient in TypeScript than I, so I'm hoping for a review from them before I dip into #5580.

**Comment by ntwb** [comment]
Looking great, thanks for the enormous amount of work everyone has been contributing 💯🙇🏼‍♂️

Also, #5585 reminds me, maybe when the release tweet goes out, we should also send some tweet or tweets thanking the sponsors and backers too.

**Comment by adalinesimonian** [comment]
I'm very close to the finish line over at vscode-stylelint! In stylelint/vscode-stylelint#265 I've rearchitected the extension from the ground up in a way that lets us test it far more easily and more thoroughly, as well as making it way easier to maintain and improve. The biggest part of this change is mostly done now, and most of what's left is to write several test suites and take a look at the one file I haven't touched yet, which is a really short one.

This PR won't fix all that in one go, but it's the biggest step with the biggest change. Further work to increase test coverage and have more end-to-end tests can happen post-v1 release. It's morning here where I live so I have plenty of time I plan to spend today getting this PR ready to merge. Shouldn't take that long now!

**Comment by jeddy3** [comment]
I've added the following to the release sequence in https://github.com/stylelint/stylelint/issues/5205#issue-837034702:

- [ ] stylelint-scss release https://github.com/stylelint-scss/stylelint-scss/pull/547
- [ ] stylelint-config-recommended-scss release https://github.com/stylelint-scss/stylelint-config-recommended-scss/pull/84
- [ ] stylelint-config-standard-scss release https://github.com/stylelint-scss/stylelint-config-standard-scss/pull/1

This is a one-time change to the sequence to improve the migration experience for SCSS users.

**Comment by ntwb** [comment]
I've commented on 2 of those 3 https://github.com/stylelint-scss PRs to switch back from GitHub to Node.js for the Stylelint version in `package.json`:

```diff
-	 "stylelint": "github:stylelint/stylelint#v14"
+    "stylelint": "^14.0.0"
```

**Comment by adalinesimonian** [comment]
vscode-stylelint 1.0.0 has been released! 🚀🎉

[GitHub Release](https://github.com/stylelint/vscode-stylelint/releases/tag/v1.0.0) | [VS Marketplace](https://marketplace.visualstudio.com/items?itemName=stylelint.vscode-stylelint)

**Comment by Mouvedia** [comment]
You forgot #5173. Will it have to wait version 15?

**Comment by jeddy3** [comment]
> You forgot https://github.com/stylelint/stylelint/issues/5173. Will it have to wait version 15?

I can be included in a minor release due to our [semver policy](https://stylelint.io/about/semantic-versioning#minor-release):

> Might break your lint build: a bug fix in a rule that results in Stylelint reporting more errors
